nomad/state/state_store.go

```diff
@@ -3442,6 +3442,20 @@ func (s *StateStore) updateDeploymentWithAlloc(index uint64, alloc, existing *st
 	state.HealthyAllocs += healthy
 	state.UnhealthyAllocs += unhealthy
 
+	// Ensure PlacedCanaries accurately reflects the alloc canary status
+	if alloc.DeploymentStatus != nil && alloc.DeploymentStatus.Canary {
```

Comment:
Actually, do we need to update `UpdateDeploymentPromotion` so that canaries that get promoted are removed from PlacedCanaries?

Reply:
I don't think so, since `UpdateDeploymentPromotion` deals with promoting the deployment/canaries I think that the values for Placed shouldn't ever be decremented. If you want to see the deployment status of a previous deployment the value for Placed should be the total that were placed

Reply:
Umm.. I'm a bit concerned about the reconciler logic.  Currently, `handleGroupCanaries` returns `canaries`, by checking `PlacedCanaries` exclusively.  If a canary got promoted, how do we ensure that scheduler no longer treats it as one.

Reply:
I believe (and can followup / confirm) Canary promotion only happens when a deployment is promoted. If there is an old deployment with non-promoted canaries, or the deployment is failed, it will use PlacedCanaries that haven't been promoted to be stopped.
```
	// Cancel any non-promoted canaries from the older deployment
	if a.oldDeployment != nil {
		for _, s := range a.oldDeployment.TaskGroups {
			if !s.Promoted {
				stop = append(stop, s.PlacedCanaries...)
			}
		}
	}

	// Cancel any non-promoted canaries from a failed deployment
	if a.deployment != nil && a.deployment.Status == structs.DeploymentStatusFailed {
		for _, s := range a.deployment.TaskGroups {
			if !s.Promoted {
				stop = append(stop, s.PlacedCanaries...)
			}
		}
	}
```
And then after the deployment is promoted, the promotion eval includes the promoted deployment canaries as canaries, but then further down `canaryState` will be false
```
	canaryState := dstate != nil && dstate.DesiredCanaries != 0 && !dstate.Promoted
```
Which is used in computeStop to stop the previous version non canary allocs. I think after all of this, the deploy is marked as successful, and even though placed canaries stays populated, we check deployment status and no new canaries are ever placed until a new deployment